blockstack/blockstack-app
# Refactor wallet's key management

> Try out this version of the Hiro Wallet - download [extension builds](https://github.com/hirosystems/stacks-wallet-web/actions/runs/1894331252).

This PR puts forward a new way to manage keys in the wallet. The motivation for these changes comes from the need to support different wallet types, such as Ledger hardware wallets. Other features we'll need in future also benefit from these changes: such as introducing a BTC wallet, or supporting multiple keys.

## Motivation
The original implementation used a “vault reducer” in the background script, where all wallet CRUD operations took place via messaging to the background. On each message, the entire in memory wallet store would be set to various atoms. Everytime a new popup script initiated, it would request `getWallet` and restore the wallet to the home page, if a session still existed.

The approach used here moves these operations out of the background script. As the plaintext key is always returned back, there's no security advantage to isolating the operations there, but does add the overhead of having to communicate via messages. The only need to use the background script was owing to the way it persists between popup script sessions. With this PR, we only leverage the persisted nature of the background script where needed, such as restoring sessions/in memory keys, but manage all other wallet operations in the popup script.

In future, I think we should explore a way in the plaintext key never reaches the popup script, and is isolated only the background's context. Would love ideas/feedback, discussion here → https://github.com/hirosystems/stacks-wallet-web/discussions/2111

## Refactor

### Multi-wallet support

Similarly to the desktop wallet, we leverage TypeScript's discriminating unions to have strongly typed wallets, using `type` value.

```ts
interface SoftwareWallet { type: 'software';  ... }
interface LedgerWallet { type: 'ledger'; ... }
type Wallet = SoftwareWallet | LedgerWallet;
```

This requires the wallet state to be represented as an object, rather than a collection of separate atoms. The atoms themselves could've been refactored to achieve this, though as explained below, we didn't take this route.

### Deriving accounts

Much of the current state utilises chains of atoms to derive new state, so that a least amount of state needs to be persisted. A good pattern to follow. However, as accounts were generated in the background script, these were cached independently. Consider the code below: setting an accounts atom shouldn't be necessary. They can be derived by the secret key.

https://github.com/hirosystems/stacks-wallet-web/blob/3ad055d795d54202c908f49dfa7e219661ab19d7/src/app/store/wallet/wallet.hooks.ts#L117-L118

Now, we derive the accounts. For this we need the secret key, and the highest generated account index. Generating accounts can be an expensive operation, though. So to both derive the accounts, and avoid the performance hit, we keep this logic in the background as a memoized function, exposed via messaging. See:

https://github.com/hirosystems/stacks-wallet-web/blob/086e7566e95c035668ab75a773a02c178bb471d0/src/background/message-handler.ts#L41-L46

### Explicitly-defined state mutations

Perhaps controversially, this PR (re)introduces Redux to manage the key state (785142b550c365a60b8ea09f35c21b84a97afe89 🥲). For the time being, we'll pay the price of having two libraries, however this a small cost. Jotai has a Redux integration which has prevented the need to change the majority of existing atoms. I've no particular attachment to Redux, but as discussed, we need a mature tool, with established patterns, ample resources, good documentation, an ecosystem of tooling, a stable API, a debugger, and most importantly, a way to be explicit about how state mutates.

In doing this, I've also redesigned the state tree to decouple the keys from their chain specific uses. Keys are now indexed using the entity pattern. Currently, with only a single wallet of id `default`. This key is then referenced by id `{ chains: { stx: { default: { ...stxSpecificDetails } } } }`.

An additional dev server to handle actions runs on port `8000`. You can connect to it remotely. Right click on Redux Devtools Extension > Open remote devtools.

### State persistence

Previously we used a handful of localStorage reads throughout the code, both directly and with `atomWithStorage`. `redux-persist` is now used to persist and rehydrate state. `migrateVaultReducerStoreToNewStateStructure` reads these localStorage values and put them into the new state structure. I recommend the [Storage Area Explorer](https://chrome.google.com/webstore/detail/storage-area-explorer/ocfjjjjhkpapocigimmppepjgfdecjkb?hl=en) extension to debug `chrome.storage`, though most the time you'll be able to examine this state with Redux devtools.

## QA

This PR will need a full regression test, as it touches almost all core wallet functionality. In addition, we should pay close attention to how the migration step works.

Existing wallet installs persist two core values in `localStorage`. Visible here _Devtools > Application > Local Storage_. `stacks-wallet-salt` and `stacks-wallet-encrypted-key`. When the new version of the wallet loads, it checks for the existence of these two values. If found, the wallet runs the migration step whereby these two values are moved into the persisted store, using `chrome.storage`. The two values from localStorage are then deleted.

The expected result of this upgrade is that the user can unlock the wallet and continue using it, unaware the step has taken place. 

### Simulating upgrade scenario

1. Download the latest build of the wallet & the build from this PR
2. Rename the existing extension's download folder to `refactor-test`
3. Install the extension with the _Load unpacked extension_ tool
4. Create a wallet, verify you can see the two localStorage values
5. Close browser
6. Delete the old extension named `refactor-test`
7. In the same folder, place the new build from this PR
8. Rename that `refactor-test`
9. Open browser (browser loads extension from same folder, simulating upgrade)
10. Test wallet works as expected

**Todo**
- [x] Fix secret key entry form
- [x] Use `chrome.storage`
- [x] Fix magic recovery code functionality
